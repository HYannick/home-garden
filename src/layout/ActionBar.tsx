import React, { MouseEventHandler, ReactNode } from 'react';
import styled from '@emotion/styled';
import { RouteComponentProps, withRouter } from 'react-router-dom';
import ArrowBack from '../core/svg/ArrowBack';
import Typography from '../components/Typography';

export interface ActionProps {
  key: number,
  icon: React.FC<{ fill?: string, stroke?: string }>,
  onClick: MouseEventHandler,
}

export interface ActionBarProps extends RouteComponentProps {
  title?: string,
  actions?: ActionProps[],
  children?: ReactNode
}

const ActionWrapper = styled('div')`
  display: flex;
  align-items: center;
  padding: 2rem 2rem 2rem 0;
`;

const BackButton = styled('div')`
  width: 5rem;
  height: 2rem;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  svg {
    width: 2rem;
    height: 2rem;
  }
`;

const Title = styled('div')`
  flex: 1;
  padding-left: 2rem;
`;

const Actions = styled('div')`
  display: flex;
`;

const Action = styled('button')`
  width: 4rem;
  height: 4rem;
  cursor: pointer;
  background: transparent;
  padding: 0;
  outline: ${({ theme }) => theme.palette.light};
  margin-left: 1.5rem;
  border: none;
  border-radius: 1rem;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  transition: 0.1s;
  svg {
    width: 2.5rem;
    height: 2.5rem;
    path {
      fill: ${({ theme }) => theme.palette.grey.darker};
      stroke:  ${({ theme }) => theme.palette.grey.darker};
      stroke-width: 0.5;
    } 
  }
`;

const ActionBar: React.FC<ActionBarProps> = ({ history, title, actions }) => {
  const goBack = () => {
    history.goBack();
  };
  return (
    <ActionWrapper>
      <BackButton onClick={goBack}>
        <ArrowBack/>
      </BackButton>
      <Title>
        <Typography noMargin variant="title" weight="800">{title}</Typography>
      </Title>
      <Actions>
        {actions && actions.map(({ key, icon: Icon, onClick}: ActionProps) => (
          <Action key={key} onClick={onClick}>
            <Icon/>
          </Action>
        ))}
      </Actions>
    </ActionWrapper>
  );
};

export default withRouter(ActionBar);
