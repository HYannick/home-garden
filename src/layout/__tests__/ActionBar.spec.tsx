import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { renderWithTheme } from '../../theme-wrapper';
import ActionBar from '../ActionBar';

afterEach(cleanup);
describe('ActionBar', () => {
  const routerProps = {
    location: { pathname: '' },
    match: jest.fn(),
    history: {},
  };


  it('should render the actionBar properly without action', () => {
    const { container } = renderWithTheme(
      <Router>
        <ActionBar title="New Species" actions={[{
          key: 1,
          icon: () => <div>Icon Edit</div>,
          onClick: jest.fn(),
        }]} {...routerProps}/>
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the actionBar properly with the right title', () => {
    const { getByText } = renderWithTheme(
      <Router>
        <ActionBar title="New Species" actions={[]} {...routerProps}/>
      </Router>,
    );
    expect(getByText(/^New Species/)).toHaveTextContent('New Species');
  });

  it('should render the actionBar properly with actions', () => {
    const actions = [{
      key: 1,
      icon: () => <div>Icon Edit</div>,
      onClick: jest.fn(),
    }];
    const { getByText } = renderWithTheme(
      <Router>
        <ActionBar title="New Species" actions={actions} {...routerProps}/>
      </Router>,
    );
    const action = getByText(/^Icon Edit/);
    expect(action).toHaveTextContent('Icon Edit');
    fireEvent.click(action);
    expect(actions[0].onClick).toHaveBeenCalledTimes(1);
  });
});
