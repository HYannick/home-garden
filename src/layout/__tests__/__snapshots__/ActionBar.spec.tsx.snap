// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionBar should render the actionBar properly without action 1`] = `
<div>
  <div
    class="css-11xigr1"
  >
    <div
      class="css-10wldv8"
    >
      <svg
        viewBox="0 0 23.225 21.762"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.342 19.647L2.115 10.881m0 0l10.227-8.766m-8.617 9.032h18"
          fill="#4A4A4A"
          stroke="#4A4A4A"
          stroke-linecap="round"
          stroke-width="3"
        />
      </svg>
    </div>
    <div
      class="css-196st0l"
    >
      <span
        class="css-zrqali"
      >
        New Species
      </span>
    </div>
    <div
      class="css-ov1ktg"
    >
      <button
        class="css-1ftdis5"
      >
        <div>
          Icon Edit
        </div>
      </button>
    </div>
  </div>
</div>
`;
